### DISCUSSION TITLE
Problems related with implicit hydrogen in SMARTS reaction.

### DISCUSSION THREAD

**Discussion by jungsdao** [rdkit/rdkit#4158] [Q&A] [answered]
**Configuration:**

- RDKit Version: 2018.09.1
- Operating system: Ubuntu 18.04.4 LTS 
- Are you using conda? Yes
- If you are using conda, which channel did you install the rdkit from? conda-forge

**Description:**

I'm trying to describe reaction involving radicals using SMARTS. 
One of the reactant is radical and product should be radical as well. 
However, when I do RunReactants, I get a closed shell molecule. 
`Chem.MolFromSmiles('C=C(CC(CC(CC(CC(CCc1ccccc1)c1ccccc1)c1ccccc1)c1ccccc1)c1ccccc1)c1ccccc1')`
Is there anything  wrong with my SMARTS string?
There had been no problem with other kinds of reaction with radical but it doesn't work for this problem. 
Any hints or comment would be appreciated. 

```
[rxn = rdChemReactions.ReactionFromSmarts('[CH2:1]-[CH1&v3:2].[CH2:3]=[C:4]>>[CH2:1]-[CH1:2]-[CH2:3]-[CH0&v3:4]')
rxn

ps = rxn.RunReactants((Chem.MolFromSmiles('C=C(CC(C[CH]c1ccccc1)c1ccccc1)c1ccccc1'),
                       Chem.MolFromSmiles('C=C(CC(CCc1ccccc1)c1ccccc1)c1ccccc1')))
print("Num products : ", len(ps))
Chem.MolToSmiles(ps[0][0])
#uniq = set([Chem.MolToSmiles(x[0],isomericSmiles=True) for x in ps])
#print(uniq)
print(Descriptors.NumRadicalElectrons(ps[0][0]))
#print("Unique products:",sorted(uniq))
Chem.MolToSmiles(ps[0][0])]
```

**Comment by greglandrum** [comment, marked as answer]
@jungsdao : the products of reactions have not been sanitized, which means that the presence/absence of radicals has not yet been re-calculated.
Try doing `Chem.SanitizeMol(ps[0][0])` before you generate the SMILES.

**Comment by jingxingzhi** [comment]
> @ jungsdao：尚未对反应产物进行消毒，这意味着尚未重新计算自由基的存在/不存在。生成SMILES之前先
> 尝试做`Chem.SanitizeMol(ps[0][0])`。

same issues, how to express atom without H by smarts(some times was free radicals)
[CH0], [OH0]，[NH0] not works

**Reply by greglandrum** [reply]
@jingxingzhi without more information about what you're having problems with I can't help.
The `H0` query definitely does work from SMARTS:
```
In [3]: m = Chem.MolFromSmiles('CC(F)(Cl)Br')

In [4]: m.GetSubstructMatches(Chem.MolFromSmarts('[#6H0]'))
Out[4]: ((1,),)

In [5]: m.GetSubstructMatches(Chem.MolFromSmarts('[CH0]'))
Out[5]: ((1,),)

In [6]: m2 = Chem.MolFromSmiles('C[C](F)Cl')

In [7]: m2.GetSubstructMatches(Chem.MolFromSmarts('[CH0]'))
Out[7]: ((1,),)

In [8]: m2.GetSubstructMatches(Chem.MolFromSmarts('[#6H0]'))
Out[8]: ((1,),)
```

**Reply by jingxingzhi** [reply]
I'm just curious why the previous conversion method didn't work，so I read the SmilesWrite.cpp and the SmartsWrite.cpp. 
I find in SmilesWrite.cpp there was "needsBracket", while SmartsWrite.cpp does not has that. 
If the needsBracket Atom was incorrectly labeled as "AtomHasImplicitH"or"AtomHCount"  ?
```
<!--StartFragment-->if (needsBracket) {    res += "[";  }
  if (isotope &amp;&amp; (isomericSmiles || (atom-&gt;hasOwningMol() &amp;&amp;                                     atom-&gt;getOwningMol().hasProp(                                         common_properties::_doIsoSmiles)))) {    res += std::to_string(isotope);  }  // this was originally only done for the organic subset,  // applying it to other atom-types is a fix for Issue 3152751:  if (!doKekule &amp;&amp; atom-&gt;getIsAromatic() &amp;&amp; symb[0] &gt;= 'A' &amp;&amp; symb[0] &lt;= 'Z') {    symb[0] -= ('A' - 'a');  }  res += symb;
  res += atString;
  if (needsBracket) {    unsigned int totNumHs = atom-&gt;getTotalNumHs();<!--EndFragment-->
```

`Chem.MolFromSmiles(Chem.MolToSmiles(Chem.MolFromSmarts('[#8H1]')))` 
![1621502209(1)](https://user-images.githubusercontent.com/37572298/118953390-599f3500-b98f-11eb-92fc-e06cc037c086.jpg)


`Chem.MolFromSmiles(Chem.MolToSmiles(Chem.MolFromSmarts('[#6]-[#8H0]')))`
![1621502236(1)](https://user-images.githubusercontent.com/37572298/118953309-47bd9200-b98f-11eb-840b-72fde1948e53.jpg)

You can see this issuewas only appeared when Atom radicals has 0 H,if there is a better way to deal with this problem？
I think may support he mutual transformation of  "needsBracket" and [*H0] can be help ,just like find Atom that mark as  "needsBracket" , re-count the number of H and show as [#numH0]?

That is very interesting！

**Reply by greglandrum** [reply]
As I said in the previous discussion: molecules which have been constructed from SMARTS are queries and do not have all the chemistry information associated with them that a molecule constructed from SMILES would have. Trying to treat a molecule constructed from SMARTS like one that was constructed from SMILES is going to fail in many, many cases.

**Reply by jingxingzhi** [reply]
Indeed, this is a very meaningful discussion. The transition between being smart and smiling seems really interesting. Maybe I have time, I will try to deal with some relevant code, to see if there are other ways to meet my needs. Also thank you for your work.
